torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1256,11 +1264,110 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
+    // TODO: Find the right and proper way to dump the debug info.
+    // We cannot print out debugging info directly.
+    LOG(ERROR) << "nccl_trace: " << dump_nccl_trace();
+  }
+}
+
+void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
+  // Logging with `FATAL`, after errMsg printed, it calls `std::abort()`
+  // to terminate the program execution.
+  LOG(FATAL) << errMsg;
+}
+
+void ProcessGroupNCCL::heartbeatMonitor() {
+  uint64_t heartBeatCounter = 0ULL;
+  while (true) {
+    // This won't have any lock since this lock is only used here.
+    // Please be aware that mutex `monitorMutex_` should not be used
+    // somewhere else to avoid the deadlock.
+    std::unique_lock<std::mutex> lock(monitorMutex_);
+    if (monitorWakeUpCV_.wait_for(
+            lock, std::chrono::seconds(heartbeatTimeoutInSec_), [&] {
+              return terminateHeartbeatMonitorThread_.load();
+            })) {
+      // For the normal complete or user interception, monitorWakeUpCV_
+      // will get notified, we early return and exit heartbeatMonitor.
+      return;
+    }
+
+    // Check the heart beat of watchdog thread.
+    auto heartbeat = heartbeat_;
+    if (heartbeat != heartBeatCounter) {
+      heartBeatCounter = heartbeat;
+    } else {
+      // No heartbeat increase detected and timeout.
+      break;
+    }
+  }
+
+  // In the timeout case and we now only dump the flight recorder
+  // to std::out. Down the road, if we have more complicated or blocking
+  // operations, we might need to use a side thread to do it.
+  dump_debugging_info();
+
+  // Create a error message reported from MonitorThread, so
+  // we throw exception and make the whole process to be killed.
+  const auto exitMsg = c10::str(
+      "[Rank ",
+      rank_,
+      "] NCCL monitor thread timeout. Basically, this could ",
+      "be due to CUDA or NCCL calls being unexpectedly blocking, ",
+      "especially when your program enters a deadlock state in watchdog"
+      "or destructors. If you see this error, please file a bug to pytorch.");
+
+  // There are two possible cases for the watchdog thread exit:
+  // Case one: desync report runs quickly, and it follows the step:
+  // collective timeout -> desync -> exception handling -> destructors
+  // -> set terminateHeartbeatMonitorThread_ -> notify monitorWakeUpCV_.
+  // So the code either early returns above or will skip the sleep below.
+  // Case two: desync might be slow or get stuck. Or we get stuck in
+  // destructors, we will sleep for some time before calling std::abort() to
+  // kill the whole process.
+  if ((terminateProcessGroup_.load() || collectiveDebugInfoMode_.load()) &&
+      !terminateHeartbeatMonitorThread_.load()) {
+    const auto logMsg = c10::str(
+        "[Rank ",
+        rank_,
+        "] NCCL watchdog is collecting debug info or process is being terminated",
+        "heartbeat monitor thread sleep for some time before abort the process.");
+    LOG(INFO) << logMsg;
```

Comment:
With this log, we can differentiate between "watchdog being stuck" and "stuck during the shutdown or desync report".

Reply:
by putting the LOG here before the sleep, its possible we print it in a case when nothing bad actually happens.  This is more of a spammy case.

IMO this should be a warning, but, it should only be printed if the heartbeat monitor actually killed the process.

case 1: heartbeat monitor kills during regular heartbeat monitor loop
case 2: heartbeat monitor kills during shutdown period
case 3: ncclwatchdog aborts, heartbeat monitor not involved
case 4: program exits cleanly

what i'm proposing is that only in case 1 and 2, we issue a warning.  But there is enough information in the warning that we can tell whether its case 1 or case 2.

for case 3, we do not want heartbeat monitor to print anything, its just noisy/distracting and not helpful.  similary for case 4.

Reply:
that's why this log can help differentiate between case 1 and case 2? I do agree we shouldn't print anything for case 3 and case 4.

First of all, for case 3 and case 4, it will not enter both if statements because `terminateHeartbeatMonitorThread_` is set to true very quickly.

For case 1, it will not hit the first if statement because watchdog somehow gets stuck, `collectiveDebugInfoMode_` will be false and `terminateProcessGroup_` is false as well.

For case 2, either `collectiveDebugInfoMode_` or `terminateProcessGroup_` will be true. That's why I added a flag here.

Both case 1 and 2, we will see an abort initiated from monitoring thread and a coredump. Does this sound make sense to you?

Reply:
Updated the how the flags work in different scenarios in the description and also I just borrowed your languages of case 1-4. So that everyone who read this PR later will understand our thinking process.

Reply:
i see, ok i think im mostly convinced.

> First of all, for case 3 and case 4, it will not enter both if statements because terminateHeartbeatMonitorThread_ is set to true very quickly.

maybe there is technically a race condition for this case though? (is it possible we print the log first and then later the  `terminateHeartbeatMonitorThread_` value is set?

Reply:
oh, that is possible but very unlikely. We set `terminateHeartbeatMonitorThread_ ` before monitorWakeUpCV_.notify_one(); And we will sleep for two minutes. The only race condition might be, monitor thread happens to wake up and then it finds that no heartbeat and then hit this log.. OK, I think what I can do is to move this logging to the second if and print out the flags. Then we can avoid the race condition.